service/history/nDCHistoryReplicator.go

```diff
@@ -840,33 +844,103 @@ func (r *nDCHistoryReplicatorImpl) backfillHistory(
 		runID,
 		lastEventID,
 		lastEventVersion))
+	localHistoryIterator := collection.NewPagingIterator(r.getHistoryFromLocalPaginationFn(
+		ctx,
+		branchToken,
+		lastEventID,
+	))
 
 	var lastHistoryBatch *commonpb.DataBlob
 	prevTxnID := common.EmptyVersion
-	for historyIterator.HasNext() {
-		historyBlob, err := historyIterator.Next()
+	historyBranch, err := serialization.HistoryBranchFromBlob(branchToken, enumspb.ENCODING_TYPE_PROTO3.String())
+	if err != nil {
+		return nil, err
+	}
+	latestBranchID := historyBranch.GetBranchId()
+	var prevBranchID string
+
+	sortedAncestors := copyAncestors(historyBranch.GetAncestors())
```

Comment:
Why do you copy and then substitute original `historyBranch` instead of just sorting ancestors in place and then create new `filteredHistoryBranch` and fill in the loop bellow?

Reply:
This is easier to achieve. For example, there are 3 branches. B1 has no ancestor. B2 has B1 ancestor, B3 has B1,B2 ancestors. 
B1
   \
    B2
  /
B3

If create new filteredHistoryBranch in the loop, then I need to fill in the previous ancestors.